### ISSUE TITLE
In version 1.4, video consumer weak interface ptrs are lost at random by the video track

### ISSUE THREAD

**Issue by adrian-j-programmer** [millicast/millicast-player-unreal-engine-plugin#30]
Hi, 

We have run into a pretty severe issue we are not sure how to solve. When a stream is being played, the video track instance can lose some video consumers, seemingly at random. We confirmed all video consumer objects are in a healthy state, and yet their weak interface ptr gets cleared in the video track. There's no feedback that it's happened, so it's tricky to detect when it happened and then repair this state. 

This happens both in PIE and standalone. It's quite random, but can be reproduced pretty very easily. It's actually connected to the crashfix for removing the consumer objects that we raised not that long ago. That crash was a result of the video track clearing its consumer objects inside the loop while iterating. But it was never investigated why these consumers are missing in the first place. 

We are more than happy to cooperate with you on finding a way to fix this, it is very important to our current business.

Thanks,
Adrian Jakubiak, Gameplay Director, Odyssey

**Comment by mcottontensor** [comment]
Looking at the code the only way the consumers are removed is if the weak pointer fails to acquire a shared pointer to the resource. If you're sure the consumers pointers are still valid then I'm unsure about what is going on.

Are you able to provide a minimal sample that exhibits this behaviour or expose some code?

**Comment by adrian-j-programmer** [comment]
Hey, we have provided a test project to David, so you can use that. Or you can set this up yourself, this issue occurs even in barebones testing levels. 
One clarification from our side - we only test our app in multiplayer client mode, never in standalone. This might have an effect on this issue.

**Comment by adrian-j-programmer** [comment]
I updated the description here, because we discovered this same thing can happen to audio consumers. We have reports of sound dropping for users while the video continues to play.

**Comment by dbaldassi** [comment]
I did not reproduced the video consumers being randomly removed. 
 Could they have been garbage collected in your game ?
 How long do you usually run the game before it happens ?

**Comment by adrian-j-programmer** [comment]
It varies a lot, but it can be anything from several minutes to half an hour. It is random. There are also rare occasions where consumers don't receive any data right from the start. 
We first noticed this in the most basic implementation, without any custom code to handle Millicast other than the barebones implementation to make it work.

**Comment by dbaldassi** [comment]
Seems I manage to have the issue, running the game for about 2hours with 4 players

**Comment by adrian-j-programmer** [comment]
I looked into the implementation of media tracks, and I think you might be right about potential threading issues in this implementation. Specifically, I believe the callbacks for creating video and audio tracks are not implemented the same way. Audio track is always created on the game thread, but the video track is not. Then, none of the callbacks that handle incoming stream data are directed to any specific thread either. 
Generally, all gameplay classes and their delegates should be stored/run on the game thread as far as I know. Perhaps directing all gameplay related code to the game thread is the path forward here.

**Comment by dbaldassi** [comment]
I think your intuition about the game thread was right.
I am now creating the VideoTrack on the game thread and pulling audio data on the game thread also 
```cpp
AsyncTask(ENamedThreads::GameThread, [this]() { PullAudioData(); });
// PullAudioData();
```
And I have no more invalid audio consumer when running the game now (ran it for ~20mns).
```cpp
PeerConnection->OnVideoTrack = [this](const std::string& Mid, RtcTrack Track) {
		AsyncTask(ENamedThreads::GameThread, [this, Mid, Track]() {
			auto videoTrack = NewObject<UMillicastVideoTrackImpl>();
			videoTrack->Initialize(Mid.c_str(), Track);
			videoTrack->AddToRoot();

			OnVideoTrack.Broadcast(videoTrack);
			{
				FScopeLock Lock(&VideoTracksCriticalSection);
				VideoTracks.Add(videoTrack);
			}
		});
	};
```

**Comment by dbaldassi** [comment]
I may need to add that I had only the issue for AudioConsumers, it did not happen for VideoConsumers for now.

**Comment by adrian-j-programmer** [comment]
I'm a bit confused. The changes you posted above are to the video consumer, not the audio consumer. Did you mean to say that you never run into this issue with audio consumers?

A couple more pieces of feedback I would give here to these changes are:
* I don't think you need the critical section for adding a video track. If you direct all these calls to the game thread, then multiple potential tasks will be run synchronously one after another on the game thread. 
* The AddToRoot function is very risky. It essentially means that the memory of these objects will never be freed, unless RemoveFromRoot is called on them explicitly. I don't think this function is ever called anywhere in the current implementation.

**Comment by dbaldassi** [comment]
The first change is for the audio not the video.

I just have added the AsyncTask on the game thread for the creation of the video track object to be consistent with the creation of the audio track.

For the AddToRoot function, it was already the case for the AudioTrack, because it was released when the garbage collector was passing, resulting in no more audio being played. When the client unsubscribe, you can see I'm calling the RemoveFromRoot function, the audio track can be released.

**Comment by adrian-j-programmer** [comment]
I actually can't find that code anywhere. I don't think the track UObjects are even cached anywhere. Anyways, that's just me sharing some thoughts on the code I see. It's great to hear that you managed to make progress on the the audio consumers being lost. Video consumers being lost was by far the more frequent bug in our experience, but at least we have a workaround for that. 
With the video consumers, could it be that the track's `OnFrame` can be called from the signalling or some internal webRTC tread? Perhaps the implementation of this function should be moved to the game thread as well without any critical sections? I will actually test this now and see which thread runs this callback in cases where the video consumer weakptr is invalid.

**Comment by dbaldassi** [comment]
Yes the ``OnFrame`` is called from a webrtc thread, but this is done in libwebrtc code so modifying means modifying unreal engine code, which is not very portable. But I can propose putting the async task directly in the OnFrame method like this : 
```cpp
AsyncTask(ENamedThreads::GameThread, [this, VideoFrame]() {
		constexpr auto WEBRTC_PIXEL_FORMAT = webrtc::VideoType::kARGB;
		
		FScopeLock Lock(&CriticalSection);

		uint32_t Size = webrtc::CalcBufferSize(WEBRTC_PIXEL_FORMAT,
			VideoFrame.width(),
			VideoFrame.height());

		if (Size != Buffer.Num())
		{
			Buffer.Empty();
			Buffer.AddZeroed(Size);
		}

		webrtc::ConvertFromI420(VideoFrame, WEBRTC_PIXEL_FORMAT, 0, Buffer.GetData());
		TArray<TWeakInterfacePtr<IMillicastVideoConsumer>> removals;  //Track the AudioConsumers to clean up

		for (auto& consumer : VideoConsumers)
		{
			if (auto c = consumer.Get())
			{
				c->OnFrame(Buffer, VideoFrame.width(), VideoFrame.height());
			}
			else
			{
				removals.Add(consumer);
			}
		}

		for (auto& consumer : removals)
		{
			VideoConsumers.Remove(consumer);
		}
	});
```

**Comment by adrian-j-programmer** [comment]
Yup, that looks like a better solution, definitely way more clear on what's going on and where. And again, this living in the game thread would mean no critical sections are needed anymore.

**Comment by dbaldassi** [comment]
ok I will remove the ``CriticalSection``, and make a push. Let me know if it fix the issue on your side.

**Comment by dbaldassi** [comment]
It is now pushed in the UE4 branch

**Comment by dbaldassi** [comment]
the main branch is for UE5, I thought you were using UE4 ?

**Comment by adrian-j-programmer** [comment]
We are in the process of migrating. It's great to have the fix in the UE4 branch for a potential hotfix release, though.

**Comment by dbaldassi** [comment]
I cherry-picked it in the dev branch.

**Comment by adrian-j-programmer** [comment]
Do you have plans to integrate this to the 5.1 branch? It's a bit of a pain now having to manually reconcile changes from both 5.1 and the dev branch.

**Comment by dbaldassi** [comment]
It is in the UE5.1 branch.
And I just merged the UE5.1 into dev and created a UE5 branch for 5.0 support

**Comment by dbaldassi** [comment]
@adrian-j-programmer can you confirm this is solved for you ?

**Comment by adrian-j-programmer** [comment]
This seems to be fixed. We will need to perform some more tests at a scale, but so far after 3 hours of testing myself I did not hit a single instance of neither video nor audio dropping out.
